opam export: record vcs commit hash when a vcs url is specified

hannesm opened the pull request:
this is useful for having an export file which can be imported and will contain the exact same sources (related to #4040), part of reproducibilty efforts

rjbou commented:
Agree to integrate this sspecification, to be sure to rebuild the exact same switch. There is a side effect: I change pinned packages (and some from repo acceptiong allowing vcs urls): e;g if you pin a package to a branch, after an export & import this package will be pinned to an exact commit. There is a way to retrieve the commit branch (it's implemented in opam or a plugin), but we loose the information if first user wanted to pin to a commit or a branch.

hannesm commented on src/client/opamSwitchCommand.ml:
```diff
@@ -477,43 +477,98 @@ let import_t ?ask importfile t =
+            let url' = OpamFile.URL.url url in
+            match url'.backend with
+            | `git ->
+              begin match OpamProcess.Job.run (OpamGit.VCS.revision (sd nv)) with
```
there I'd get an `OpamPackage.Version.t`, where I don't know how to extract the commit hash

rjbou replied:
Simply get the string `OpamVersion.to_string`. There is [no conversion](https://github.com/ocaml/opam/blob/master/src/format/opamPackage.ml#L24) step, the only thing is that `OpamVersion.of_string` checks that it is an alphanum, plus some punctuation.
The hash is converted to version  [here](https://github.com/ocaml/opam/blob/master/src/repository/opamVCS.ml#L99).

rjbou replied:
@AltGr shouldn't we change `{VCS,OpamRepository}.revision` type to a `string option` ?
There is only some calls and they are all directly used as a string:
```
src/client/opamAction.ml:479:          let rev = OpamProcess.Job.run (OpamRepository.revision src u) in
src/client/opamAction.ml-480-          Printf.sprintf "pinned(%s%s)"
src/client/opamAction.ml-481-            (OpamUrl.to_string u)
src/client/opamAction.ml-482-            (OpamStd.Option.to_string
src/client/opamAction.ml-483-               (fun r -> "#"^OpamPackage.Version.to_string r) rev)
--
src/client/opamAdminCommand.ml-94-        let module VCS = (val OpamRepository.find_backend_by_kind vcs) in
src/client/opamAdminCommand.ml:95:        match OpamProcess.Job.run (VCS.revision repo_root) with
src/client/opamAdminCommand.ml-96-        | None -> date ()
src/client/opamAdminCommand.ml-97-        | Some hash -> OpamPackage.Version.to_string hash
--
src/client/opamListCommand.ml-557-      OpamSwitchState.url st nv >>| OpamFile.URL.url >>= fun url ->
src/client/opamListCommand.ml-558-      OpamSwitchState.source_dir st nv |>
src/client/opamListCommand.ml-559-      OpamFilename.opt_dir >>= fun srcdir ->
src/client/opamListCommand.ml:560:      OpamProcess.Job.run (OpamRepository.revision srcdir url) >>|
src/client/opamListCommand.ml-561-      OpamPackage.Version.to_string
src/client/opamListCommand.ml-562-    in
```

rjbou replied:
After discussion, `OpamRepository.revision` changes its type to `string`  (on a  further PR).

hannesm commented on src/client/opamSwitchCommand.ml:
```diff
@@ -477,43 +477,98 @@ let import_t ?ask importfile t =
   t
 
-let read_overlays (read: package -> OpamFile.OPAM.t option) packages =
+let read_overlays sd ~repos_roots (read: package -> OpamFile.OPAM.t option) packages =
```
done in 77478b24

hannesm commented on src/client/opamSwitchCommand.ml:
```diff
@@ -477,43 +477,98 @@ let import_t ?ask importfile t =
+let export rt ?(full=false) ?(switch=OpamStateConfig.get_switch ()) filename =
+  let gt = rt.repos_global in
   let root = OpamStateConfig.(!r.root_dir) in
+  OpamSwitchState.with_ `Lock_none ~rt ~switch gt @@ fun st ->
```
done in 77478b24

hannesm commented:
I rebased on top of master and added a commit which addresses the review comments. when merging #4040 (that now adds an optional labeled argument `freeze` to `OpamSwitchCommands.export`) the code from this PR should only be executed if `freeze` is provided and true.

hannesm commented:
now that #4040 is merged, I rebased this on top of master - and by doing this I squashed all commits (which simplified the rebase). I also guarded, as discussed some weeks ago, this behaviour behind an optional labeled `freeze` argument to `export`.

apart from this PR, only df3bd358cc3865090f970372f4683e8b943b0515 and 365ac0274a7dc287c97a0bcf4e55cff892949018 are needed for orb. I can cherry-pick them on here if that is preferable.

thanks for reviews, discussion, and merges :)

hannesm commented on src/client/opamSwitchCommand.ml:
```diff
@@ -494,17 +494,57 @@ let import_t ?ask importfile t =
+          | [] ->
+            OpamConsole.warning "Ignoring package %s, no checksum"
+              (OpamPackage.Name.to_string nv.name);
+            None
```
I'm a bit uneasy about this and the failure case below -- now they emit a `OpamConsole.warning`, but maybe the export should fail hard (this way, we could guarantee once frozen and full export is done, this can be imported). But: I don't know how to fail hard in this code (`export` returns a `unit`, thus there must be some exception mechanism or a direct `exit`?)

rjbou commented:
I updated few things:
* add freeze (that implies full) as a `opam switch export` option
* changes warnings into errors
* use [Opamrepository.revision] to handle other VCSs
* use [OpamFile.URL.with_url] to not recreate an url
* fix source dir look up
@hannes, can you check that this work on your tests?

hannesm commented:
thanks @rjbou -- indeed this works fine for me. tested with:
```bash
$ opam pin add orb https://github.com/hannesm/orb.git#active # this includes this branch and the two commits mentioned above on the opam side
$ opam repo add test-unikernels https://github.com/roburio/reproducible-unikernel-repo.git
$ orb orb build --twice --repo=test-unikernels,default unikernel-hello-hvt
```

and then the resulting /tmp/bi-unikernel-hello*/*opam-switch contains ocamlgraph and unikernel-hello-hvt which are pinned to the respective git commits.

rjbou commented on src/client/opamCommands.ml:
```diff
@@ -2248,7 +2255,7 @@ let switch =
       OpamRepositoryState.with_ `Lock_none gt @@ fun rt ->
       OpamSwitchCommand.export rt
-        ~full
+        ~full:(full||freeze)  ~freeze
```
```suggestion
        ~full:(full || freeze) ~freeze
```

rjbou commented on src/client/opamSwitchCommand.ml:
```diff
@@ -494,18 +494,65 @@ let import_t ?ask importfile t =
+               url)
+            opam
+        in
+        opam)
```
```suggestion
        OpamFile.OPAM.with_url
          (OpamFile.URL.with_url
             { url_t with hash = Some (OpamPackage.Version.to_string hash) }
             url)
          opam)
```

rjbou commented on src/client/opamSwitchCommand.ml:
```diff
@@ -494,18 +494,65 @@ let import_t ?ask importfile t =
+       | _ -> opam)
+    | `rsync ->
+      OpamConsole.error_and_exit `Not_found
+        "%s is pinned with rsync, it can't be exported with --freeze."
```
```suggestion
        "%s is path pinned, it can't be exported with --freeze."
```

hannesm commented on src/client/opamSwitchCommand.ml:
```diff
@@ -494,18 +494,65 @@ let import_t ?ask importfile t =
+       | [] ->
+         OpamConsole.error_and_exit `Not_found
+           "%s url doesn't have an associated checksum, \
+            it can't be exported with --freeze."
```
sure, that is an option. I did not write that code, since I'm unsure how this should look like -- once here, can it be assumed that the sources are already downloaded (I suspect no), thus export would now do potentially lead to network communication.

rjbou replied:
If the cache is not cleaned, the archive can be no more locally reachable, and needs a redownload.
On this part, I made it referring to https://github.com/ocaml/opam/pull/4055#discussion_r388347604
